mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,354 @@
+        }
+        let prices_in_series = join_all(prices_futs)
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>() // if an error in any future - return the error
+            .mm_err(|err| ApiIntegrationRpcError::from_api_error(err, None))?;
+
+        let quotes = src_dst
+            .into_iter()
+            .zip(prices_in_series.iter())
+            .map(|((src, dst), series)| {
+                let dst_price = cross_prices_average(series);
+                ((src, dst), dst_price)
+            })
+            .collect::<HashMap<(_, _), _>>();
+        for q in &quotes {
+            log::debug!(
+                "query_destination_token_prices src/dst quote={:?} {}",
+                q,
+                q.1.to_decimal()
+            );
+        }
+        self.update_with_lr_prices(&quotes);
+        Ok(())
+    }
+
+    /// Estimate the needed source amount for LR swap, by dividing the known dst amount by the src/dst price
+    #[allow(clippy::result_large_err)]
+    fn estimate_source_token_amounts(&mut self) -> MmResult<(), ApiIntegrationRpcError> {
+        for lr_data in self.inner.values_mut() {
+            let Some(ref dst_price) = lr_data.lr_price else {
+                continue;
+            };
+            let dst_amount = lr_data
+                .dst_amount
+                .ok_or(ApiIntegrationRpcError::InternalError("no dst_amount".to_owned()))?;
+            let dst_amount = mm_number_from_u256(dst_amount);
+            if let Some(src_amount) = &dst_amount.checked_div(dst_price) {
+                lr_data.src_amount = Some(mm_number_to_u256(src_amount)?);
+                log::debug!(
+                    "estimate_source_token_amounts lr_data.order.coin={} dst_price={} lr_data.src_amount={:?}",
+                    lr_data.order.coin,
+                    dst_price.to_decimal(),
+                    lr_data.src_amount
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Run 1inch requests to get LR quotes to convert source tokens to tokens in orders
+    async fn run_lr_quotes(&mut self, ctx: &MmArc) -> MmResult<(), ApiIntegrationRpcError> {
+        let mut src_dst = vec![];
+        let mut quote_futs = vec![];
+        for ((src_token, dst_token), lr_data) in self.inner.iter() {
+            let Some(src_amount) = lr_data.src_amount else {
+                continue;
+            };
+            let src_contract = eth_addr_to_hex(
+                lr_data
+                    .src_contract
+                    .as_ref()
+                    .ok_or(ApiIntegrationRpcError::InternalError("no contract".to_owned()))?,
+            );
+            let dst_contract = eth_addr_to_hex(
+                lr_data
+                    .dst_contract
+                    .as_ref()
+                    .ok_or(ApiIntegrationRpcError::InternalError("no contract".to_owned()))?,
+            );
+            let chain_id = lr_data
+                .chain_id
```

Comment:
May be just skip `lr_data` if some of these values is None instead of returning error?

Reply:
My point was we should skip those lr_data for which we did not receive data from the LP provider for some token considered for LR (this may happen in reality),
but we should stop with an error if there is something wrong with our config - we must fix it before resuming swaps.

Reply:
I see. Please add doc comments for the LrData fields, clarifying which values are obtained from our configuration or other internal sources, and which come from 3rd party

Also, please specify whether the source or destination contract is missing in the InternalError message. Right now, it just says `"no contract"`.

Reply:
added more doc comments and extended error msgs